File: GoF_Design_Pattern/Iterator_Pattern/java/BookShelf.java
package Software_Engineering.GoF_Design_Pattern.Iterator_Pattern.java;


public class BookShelf implements Aggregate {

    private Book[] books;
    private int last = 0;

    public BookShelf(int maxSize) {
        this.books = new Book[maxSize];
    }

    public Book getBookAt(int index) {
        return books[index];
    }

    public void appendBook(Book book) {
        this.books[last] = book;
        last ++;
    }

    public int getLength() {
        return last;
    }

    @Override
    public Iterator iterator() {
        return new BookShelfIterator(this);
    }

}

File: GoF_Design_Pattern/Iterator_Pattern/python/BookShelf.py
#! /etc/anaconda/python3

from BookShelfIterator import BookShelfIterator
from aggrgate import Aggregate


class BookShelf(Aggregate):

    __books = []
    __last = 0

    def __init__(self, max_size):
        self.__books = [None] * max_size

    def get_book_at(self, index):
        return self.__books[index]

    def insert_book(self, book):
        self.__books[self.__last] = book
        self.__last += 1

    def get_length(self):
        return self.__last

    def iterator(cls):
        return BookShelfIterator(cls)
